algorithms/linfa-reduction/src/random_projection/sparse/algorithms.rs

```diff
@@ -0,0 +1,101 @@
+use linfa::{prelude::Records, traits::Fit, Float};
+use ndarray::Ix2;
+use ndarray_rand::rand_distr::StandardNormal;
+use rand::{distributions::Bernoulli, prelude::Distribution, thread_rng, Rng};
+use sprs::{CsMat, TriMat};
+
+use super::super::common::johnson_lindenstrauss_min_dim;
+use super::hyperparams::SparseRandomProjectionParamsInner;
+use super::{SparseRandomProjectionParams, SparseRandomProjectionValidParams};
+use crate::{impl_proj, ReductionError};
+
+/// Embedding via sparse random projection
+pub struct SparseRandomProjection<F: Float> {
+    projection: CsMat<F>,
+}
+
+impl<F, Rec, T> Fit<Rec, T, ReductionError> for SparseRandomProjectionValidParams
+where
+    F: Float,
+    Rec: Records<Elem = F>,
+    StandardNormal: Distribution<F>,
+{
+    type Object = SparseRandomProjection<F>;
+
+    fn fit(&self, dataset: &linfa::DatasetBase<Rec, T>) -> Result<Self::Object, ReductionError> {
+        let n_samples = dataset.nsamples();
+        let n_features = dataset.nfeatures();
+
+        let n_dims = match &self.params {
+            SparseRandomProjectionParamsInner::Dimension { target_dim } => *target_dim,
+            SparseRandomProjectionParamsInner::Precision { precision } => {
+                johnson_lindenstrauss_min_dim(n_samples, *precision)
+            }
+        };
+
+        let scale = (n_features as f64).sqrt();
+        let p = 1f64 / scale;
+        let dist = SparseDistribution::new(F::cast(scale), p);
+        let mut rng = thread_rng();
```

Comment:
You can pass a `Rng` as you do for `GaussianRandomProjection` in order to be able to strictly reproduce runs.

Reply:
I'm not sure what is the best way to do this.
The `fit` function of the `Transform` trait takes `&self` as parameter, so the only way to use the `Rng` is to `clone()` it. This means that using the same `GaussianProjectionParamters` struct twice in a row to create a projection of a given size will result in the **same** projection. Is this the expected behavior?

Another issue is that few algorithms in `linfa` can use a custom `Rng`, so you would only have reproducibility for the projection part.

Reply:
> Is this the expected behavior?

Yes, any algorithm using randomness should offer a way to be reproducible for debugging/testing purpose.  
For `Rng` what I have in mind was things like in `linfa-clustering`, check out `kmeans` or `gaussian mixture` for instance where the random generator is not an option.

Reply:
Updated in [3569c10](https://github.com/rust-ml/linfa/pull/332/commits/3569c1035f83e1818704eb14d582ee143268a3e4), both structs now have RNGs; this field is no longer an option, and `Xoshiro256Plus` is used by default if not provided.